### ISSUE TITLE
Pure virtual function definitions in cpp are counted as magic numbers

### ISSUE THREAD

**Issue by noricor** [metrixplusplus/metrixplusplus#36]
Hello,

first of all, thank you very much for this great tool.
I have been using it for a couple of years now and I find it extremely helpful.

Now regarding the issue at hand:

In cpp interface definitions pure virtual functions usually have the following syntax:
```
class Foo {
    virtual ~Foo();
    virtual void bar(void) = 0;
};
```
On default metrix++ counts 0 as a magic number, but in this specific context it isn't really one.
I am aware that there is a switch for magic numbers that allows ignoring 0 and 1,
but I think this specific case should already be covered on default.

I did not yet start thinking how this specific case could be solved and 
maybe this issue is already well known and you decided to keep it that way?

I am happy to hearing from you.

noricor

**Comment by avkonst** [comment]
Hi,
I think it should be possible to count magic numbers only in regions on type 'function'. It seems like it counts it for 'class' regions too.
The project is community driven. Would you like to patch the plugin code and submit pull request? The change will be about adding a condition to include only function regions in calculation and ignore the rest.

**Comment by noricor** [comment]
Hi, 

thanks for the fast response and sure, I am happy to contribute.

I just gave it a try and set the `region_type_mask` in the magic plugin to `api.Region.T.FUNCTION` 

It works for the case at hand, but it also impacts other cases.

Here are some cases I could already identify: 

```cpp
// cpp interface class
class iTest
{
    virtual ~iTest();
    virtual void method() = 0; // no magic number any longer
};

// modern cpp member initialisation
class cTest
{
    int member = 5; // no magic number any longer
    int other;
};

// bit fields
struct sTest
{
    unsigned char b1 : 3; // no magic number any longer
    unsigned char : 0; // no magic number any longer
    unsigned char b2 : 8; // no magic number any longer
};
```

Before submitting the pull request, I still need to think this over a bit more,
plus I need to figure out where to put the unit tests.

Cheers

**Comment by avkonst** [comment]
Would you say that struct/class fields initialized to numbers are magic numbers?

Tests run the tool over a set of sample codes and compare the expected output (gold files) and actual output. There is an environment variable which allows to regenerate gold files.

**Comment by noricor** [comment]
No, actually I think with this approach we can eliminate lots of other false positives, so it is good.

It is the false negatives I still have to think about, here an actual magic number
may not be recognized any longer because of this change. 

Plus the symmetry to other region types should also still be there.

I found one more case where a non-const static member initialization is also no longer recognized as a magic number
```cpp
class Foo 
{
static int bar;
};

int Foo::bar = 42; // no magic number any longer (I could live with that too)
```

Thanks for the information about the unit tests, I will give it a try.

**Comment by avkonst** [comment]
Maybe it could be made configurable with cli option? So everybody could choose what types of regions to include?

**Comment by noricor** [comment]
With the latest fix in #38 pure virtual function definitions are not considered magic numbers any more,
but other things mentioned above (e.g. bitfields) are still detected as magic numbers,
so future work may be needed here.
